### DISCUSSION TITLE
Sqlalchemy with distributed task(Celery)

### DISCUSSION THREAD

**Discussion by aviciot** [sqlalchemy/sqlalchemy#9388] [Usage Questions]
Working on a Celery based program which (hopefully) be able to execute queue of thousands (light) queries.
Each select statement returns few lines of data and store it .

I manage to reuse db connection for the celery worker - (workers running in pool=gevent with high number concurrency).
I'd like to advise for recommended approach for such operation:

1.Non async - create engine with predefine pool_size and share it to celery task 

```
engine = create_engine(
    "postgresql+psycopg2://me@localhost/mydb", pool_size=300, max_overflow=0
)

```

In function:
DBFacotry - Based class which holds reusable engine connection (self.db)

```
@app.task(bind=True ,base=DBFacotry)
def load_to_db(self, *args, **kwargs)
   ....
   source_name = kwargs.get('source_name')
    db_connection = self.db[source_name]
    conn = engine.connect()
    sql = text(sql)
    query = conn.execute(sql)
    df = pd.DataFrame(query.fetchall())
    return df
```

2.sqlalchemy - async version 
I have less experience with it , but seems it should work better as the celery task submitted in async fashion to the broker .
I guess it would be similar as sync version - meaning share the engine and then connect in the function ,
Would be thankful for some guidance for recommend approach and if async is the way from where to start.

**Comment by zzzeek** [comment]
I havent used celery directly but I would not share a pool across process boundaries.   A new subprocess should have its own connection pool and it's not safe to share a PostgreSQL connection among multiple processes or threads simultaneously.  background is at https://docs.sqlalchemy.org/en/20/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork

**Reply by aviciot** [reply]
Thanks @zzzeek  , 
I'm running the celery workers in gevent execution pool, spawning ~50-100 green threads .
each task fetching a small amount of the data and and save it somewhere  .
As those light IO tasks, I was thinking gevent suit more than running celery in prefork mode (multiprocessing ).
Tasks queue is quite serious - speaking of thousands , so it obviously it make sense to run as much as simultaneously tasks as possible .

Following your suggestion:
each task will create an engine -> run the sql --> disconnect?

```
@app.task(bind=True ,base=DBFacotry)
def load_to_db(self, *args, **kwargs)

#  poolclass=NullPool to close connection
    source_db_url = kwargs.get('source_db_url ' ,poolclass=NullPool)
    engine= create_engine(source_db_url )
    conn = engine.connect()
    query = conn.execute(text(sql))
    conn.invalidate()
    db.dispose()
     ... rest of code
```
will it not be an overhead of number of simultaneously connections ?
Second Q ,is there any advantage running such program based on sqlalchemry async engine?

**Reply by CaselIT** [reply]
You can use a single engine, but each task should call engine.connect() and then connection.close() to return it to the pool (or better yes use as context manager)
This way each task uses its own connection.

Regarding asyncio, it mostly depends on the preference, I don't think there would be a lot of difference compared to green threads. Also it depends on what db you are using, outside postgresql, the other async drivers are quite lacking

**Comment by CaselIT** [comment]
Hi,

I don't think celery support asyncio natively (at leadt this is still open https://github.com/celery/celery/issues/3884) so I'm not sure what you gain here by doing asyncio.

Leaving the sync option I think, like mike said, use an engine for each worker process. The same worker process can share the engine if it uses multiple threads / green thread, but each task should use it's own connection/session, that's not shared with other threads and is not used concurrently

Basically the usage is the same as when using a web-server: each request (task in celery) gets its own connection

**Reply by aviciot** [reply]
@CaselIT  thanks ,
I think you right ,  I also read somewhere Celery has some limitation with asyncio ,
Thee nature of this program is async : 

I have one thread that keep executing the tasks
and other one that keep monitoring tasks progress and result

So when I submit celery task - it doesn't wait and from main program 
```
    async def run_task(self):
        self.task_run = proccess_rule.delay(**task_kwargs)

```

So based on that I though sqlalchemy async would be more efficient than sync engine
